### PR TITLE
[WIP :fire:] P2C CrossChain command execution - New TX 

### PR DESCRIPTION
**Pull request by DerTiedemann**
Related EVM PR: https://github.com/chain4travel/caminoethvm/pull/74

## Why this should be merged
This adds crosschain communication from p->c. Sending funds does already exist, but that is not sufficient for what we want to do in the future - we can only send funds and cannot influence the chainstate of the c-chain from any other chain. This PR adds a harness for that functionality, prominently for setting the BaseFee but also for other usecases like setting KYC state. The commands have access to the EthVM statedb and can do what ever a precompiled smartcontract could do. 
## How this works
Lets say a proposal on P-Chain has passed and the BaseFee has to be changed on C-Chain aswell.

1. The P-Chain writes the serialzied Command to a predefined location/offset (currently based on command type, but can be revised to be more general - i did it this way to avoid conflicts if multiple commands are sent in the same block) in the shared memory of the C-Chain.
2. The C-Chain checks every block if sth was written to shared memory on those magic 🪄 addresses and deserializes the command and adds a wrapper transaction with the command attached to the mempool.
3. The transaction checks if the command attached to it is exactly the same as inside of the shared memory and executes logic based on the command. In this case set the baseFee value in the admin contract.

## How this was tested
Sent a few command tx and checked if the chain still boots and the messages arrive.
I held off writing extensive tests as i would like to have feedback on the whole idea.

### CODE REVIEW

File: vms/platformvm/txs/commands/camino_command_set_base_fee.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+)
+
+type CommandSetBaseFee struct {
+	BaseFee uint64 `serialize:"true"`
+}
+
+func (cmd *CommandSetBaseFee) Verify() error {
+
+	switch {
+	case cmd.BaseFee > 0:
```
Comment (evlekht) [inline vms/platformvm/txs/commands/camino_command_set_base_fee.go]: thats the opposite

File: vms/platformvm/txs/commands/camino_command_set_base_fee.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
+package commands
+
+import (
+	"fmt"
+	"math/big"
+
+	"github.com/ava-labs/avalanchego/snow"
+	"github.com/ava-labs/coreth/core/state"
```
Comment (evlekht) [inline vms/platformvm/txs/commands/camino_command_set_base_fee.go]: I don't think that its a good idea to make 2-way dependency between caminogo and caminoethvm (or one-way on avax-evm)

File: vms/platformvm/txs/executor/atomic_tx_executor.go [hunk 3]
```diff
@@ -90,6 +90,10 @@ func (e *AtomicTxExecutor) ExportTx(tx *txs.ExportTx) error {
 	return e.atomicTx(tx)
 }
 
+func (e *AtomicTxExecutor) SendCommandTx(tx *txs.SendCommandTx) error {
```
Comment (evlekht) [inline vms/platformvm/txs/executor/atomic_tx_executor.go]: that must go to camino executor

File: vms/platformvm/txs/executor/proposal_tx_executor.go [hunk 4]
```diff
@@ -88,6 +88,10 @@ func (*ProposalTxExecutor) ExportTx(*txs.ExportTx) error {
 	return errWrongTxType
 }
 
+func (*ProposalTxExecutor) SendCommandTx(*txs.SendCommandTx) error {
```
Comment (evlekht) [inline vms/platformvm/txs/executor/proposal_tx_executor.go]: we have dedicated file for this, see camino visitor

File: vms/platformvm/txs/executor/standard_tx_executor.go [hunk 5]
```diff
@@ -55,6 +55,10 @@ func (*StandardTxExecutor) RewardValidatorTx(*txs.RewardValidatorTx) error {
 	return errWrongTxType
 }
 
+func (*StandardTxExecutor) SendCommandTx(*txs.SendCommandTx) error {
```
Comment (evlekht) [inline vms/platformvm/txs/executor/standard_tx_executor.go]: we have dedicated file for this, see camino visitor

File: vms/platformvm/txs/executor/tx_mempool_verifier.go [hunk 6]
```diff
@@ -68,6 +68,10 @@ func (v *MempoolTxVerifier) ExportTx(tx *txs.ExportTx) error {
 	return v.standardTx(tx)
 }
 
+func (v *MempoolTxVerifier) SendCommandTx(tx *txs.SendCommandTx) error {
```
Comment (evlekht) [inline vms/platformvm/txs/executor/tx_mempool_verifier.go]: we have dedicated file for this, see camino visitor

File: vms/platformvm/txs/mempool/issuer.go [hunk 7]
```diff
@@ -74,6 +74,11 @@ func (i *issuer) TransformSubnetTx(*txs.TransformSubnetTx) error {
 	return nil
 }
 
+func (i *issuer) SendCommandTx(*txs.SendCommandTx) error {
```
Comment (evlekht) [inline vms/platformvm/txs/mempool/issuer.go]: we have dedicated file for this, see camino visitor

File: vms/platformvm/txs/mempool/remover.go [hunk 8]
```diff
@@ -76,3 +76,8 @@ func (*remover) RewardValidatorTx(*txs.RewardValidatorTx) error {
 	// this tx is never in mempool
 	return nil
 }
+
+func (*remover) SendCommandTx(*txs.SendCommandTx) error {
```
Comment (evlekht) [inline vms/platformvm/txs/mempool/remover.go]: we have dedicated file for this, see camino visitor

File: go.mod [hunk 9]
```diff
@@ -11,9 +11,11 @@ go 1.19
 require (
 	github.com/Microsoft/go-winio v0.5.2
 	github.com/NYTimes/gziphandler v1.1.1
+	github.com/ava-labs/coreth v0.12.3
```
Comment (evlekht) [inline go.mod:14]: we shouldn't depend on avax evm, imo
its not as bad as if caminogo would depend on caminoethvm, but its still bad idea to depend on avax repository

File: vms/platformvm/txs/commands/camino_command_set_base_fee.go [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+
+	switch {
+	case cmd.BaseFee > 0:
+		return fmt.Errorf("baseFee has to be greater that 0")
+	}
+
+	return nil
+}
+
+func (cmd *CommandSetBaseFee) EVMStateTransfer(ctx *snow.Context, state *state.StateDB) error {
```
Comment (evlekht) [inline vms/platformvm/txs/commands/camino_command_set_base_fee.go]: I think that its not right to have evm code in pvm
EVM stuff must be in evm, same true for any other vm

**Review by evlekht — CHANGES_REQUESTED** [review]

File: go.mod (hunk 9 above)
Comment (DerTiedemann) [reply go.mod:14]: this only pulls in a seperate package that has no relation to anything else from coreth. One could put everything into caminogo, but then we get duplicated code and codec madness. Lets discuss in a sync

**Pull request closed without merging**